src/rpc_server.rs

```diff
@@ -218,10 +218,10 @@ pub trait RPC {
     async fn next_receiving_address(key_type: KeyType) -> ReceivingAddress;
 
     /// Return all known keys, for every [KeyType]
-    async fn known_keys() -> Vec<SpendingKey>;
+    async fn known_keys() -> Vec<SpendingKeyRangeIter>;
 
     /// Return known keys for the provided [KeyType]
-    async fn known_keys_by_keytype(key_type: KeyType) -> Vec<SpendingKey>;
+    async fn known_keys_by_keytype(key_type: KeyType) -> SpendingKeyRangeIter;
```

Comment:
> IMHO iterators are inappropriate for an RPC interface. Instead, I would suggest returning a `SpendingKeyRange` and equip this new data type with mapping into `SpendingKeyRangeIter`.

Yeah it's an interesting point with a few facets to consider.    I did think about it some while implementing, but don't have strong feelings one way or another.  

> The reasons why I think iterators are inappropriate for RPC interfaces:
> 
> * Iterators are rather specific to rust. Not all languages have them (_e.g._ python) and this interface forces the other-language programmer to implement iterators.

This opens a larger discussion.   The elephant in the room is that our entire RPC mechanism is 100% tied to rust already due to tarpc.  I previously opened issues, investigated alternatives, and had some discussions around that but did not find much team appetite around changing it to something language agnostic, such as json-rpc, or devoting work to make tarpc multi-language friendly.  I think mainly there has always been higher priorities.  So I've come to accept that for launch our RPC layer will require rust clients, and that's that.

Also, if accessing the struct's fields directly, the SpendingKeyRangeIter struct becomes usable with a simple for loop, even for a non iterator language.   eg:

```
for i = iter.curr; i < iter.last i ++ {
   print iter.nth(i)
}
```

We can just think of it as a struct and sort of forget that it implements Iterator trait.

So yeah I considered this point, but my overall take is that right now we have a rust-specific RPC layer and at such time as we write a non-rust layer we can take care of any such issues. 

But yeah, I understand your point of view and I might even be the one making that argument in another scenario.  I don't feel very strongly either way.... with this draft PR I'm kinda just trying stuff out, and this seemed kind of a neat and compact way to do it.

I will give it some more thought.


> * Iterators hold state, but this state is irrelevant here.

mostly agree.

> * Iterators access data in a way that makes abstraction of where that data lives. In this case, the relevant data does not live on the other side of the RPC interface, so there is no need for the abstraction layer. 
Moreover, using iterators suggests (incorrectly) that the data does live on other side of the RPC interface.

I don't see that an iterator suggests that data lives on the other side of RPC interface.  Indeed it would be surprising if calling next() on the iterator invoked another RPC call, though I agree such things are possible and sometimes done.   But that would/should be clearly documented behavior.

-------------

Let's make our discussion a bit more concrete.

The known_keys() RPC presently returns `Vec<SpendingKeyRangeIter>`.  It is a vec because we have one itererator for each supported key-type.   (presently just symmetric and generation).

in AddressBook::render() of the dashboard, I added the following that writes out all the known keys for all supported key types:

```
        let matrix = self
            .data
            .lock()
            .unwrap()
            .iter()
            .flat_map(|iter| {
                iter.clone().map(|key| {
                    vec![
                        KeyType::from(&key).to_string(),
                        key.to_address()
                            .to_display_bech32m_abbreviated(self.network)
                            .unwrap(),
                    ]
                })
            })
            .collect_vec();
```

It's a bit verbose, but actually pretty straight-forward considering it is displaying a range of keys for every supported key-type.   The flat_map() converts the sub-lists into a single flat list.

So that's the primary use-case I was enabling with this.  Yes, it can be done other ways, eg with SpendingKey::derive_child() and a limit, but they will probably be more verbose/awkward, at least for our rust clients.

Reply:
In #278 the RPC methods return a SpendingKeyRange which is friendly for non rust languages, but also has methods for obtaining an iterator or a parallel iterator.